File: solutions/Real Contests/2023 USACO Mar/Q2/m2.cpp
#include <iostream>
#include <map>
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

int main() {
    int C, N;
    cin >> C >> N;

    vector<int> teams(N);
    for (int i = 0; i < N; i++) {
        int bitmask = 0;
        string s;
        cin >> s;
        for (int j = 0; j < C; j++) {
            if (s[j] == 'H') {
                bitmask |= (1 << j);
            }
        }
        teams[i] = bitmask;
    }

    for (int i = 0; i < N; i++) {
        int max_diff = 0;
        for (int j = 0; j < N; j++) {
            if (i != j) {
                int diff;

                diff = BitCount(teams[i] ^ teams[j]);

                max_diff = max(max_diff, diff);
            }
        }

        cout << max_diff << endl;
    }

    return 0;
}

File: USACO/Real Contests/2023 USACO Mar/Q2/m2.py
# read input values
C, N = map(int, input().split())

# read teams as a list of strings
teams = []
for i in range(N):
    teams.append(input().strip())

# function to calculate the difference between two teams
def team_difference(team1, team2):
    diff = 0
    for i in range(C):
        if team1[i] != team2[i]:
            diff += 1
    return diff

# calculate maximum difference for each team
for i in range(N):
    max_diff = 0
    for j in range(N):
        if i != j:
            diff = team_difference(teams[i], teams[j])
            max_diff = max(max_diff, diff)
    print(max_diff)
